### ISSUE TITLE
WE NEED DOM ACCESS

### ISSUE THREAD

**Issue by genesislubrigas** [MicrosoftEdge/WebView2Feedback#1152]
Why do you create new web browser control just to prevent us from using DOM access ?  What is the purpose of programming if not able to use the functionality of DOM ?

There is nothing wrong with vb.net webbrowser.  Security issue is not really directly connected with the webbrowser control.

Its not programming control.  The security issue is about something that you cannot control that is why it was removed.

Basically, I hate the webview2 control.  It doesnt justify the webbrowser control replacement.

**Comment by genesislubrigas** [comment]
You should provide documentation and example instead for the alternative.

**Comment by ukandrewc** [comment]
There is DOM access, you use it through ExecuteScriptAync, it is in the documentation.

**Comment by zydjohnHotmail** [comment]
Hello:
@ukandrewc I think if you can release nuget package for your repo:
https://github.com/ukandrewc/Webview2.Winforms.DOM
Make sure your repo match the current version of WebView2 (1.0.774.44), and provide some C# code examples, then I think it is at least a temporary work-around.
Do you have some plan for this in the near future?
By the way, I heard that MS will release Blazor Desktop, I hope the Desktop will be something like a browser or like WebView2, not a server, like Blazor server side.  If Blazor Desktop will be something like a WebView2, then having a DOM access seems to be necessary, right?  It will be much better if we can access DOM via C#, not only via JavaScript.
Please advise!

**Comment by ukandrewc** [comment]
I'm waiting until WebView2 is more stable and has a few pipelined features. I don't have the time to stay on top of every release, and I'm unlikely to release a Nuget package.

@zydjohnHotmail You can use intellisense and object browser to see the methods. It pretty much follows the JS DOM model, and the release notes give an idea of usage.

Not sure about the future, that really entirely depends on MS and if they change their mind about a .Net DOM. I will most likely update the current control with later releases, but it's only intention right now is proof of concept.

**Comment by genesislubrigas** [comment]
ok i will check

**Comment by darbid** [comment]
@genesislubrigas 
> You should provide documentation and example instead for the alternative.

There is another WebView2 DOM project in GitHub which may indicate a more open source option.

[It can be found here. WebView2.DOM](https://github.com/R2D221/WebView2.DOM)

A little bit from the ReadMe;
> With this library you can control the contents of a WebView2 directly within C#, with all the benefits of type safety that C# gives. For example, you can do this:

```
using WebView2.DOM;
using static WebView2.DOM.HTMLElementTag;

// ...

var document = window.document;

// Type-safe element creation
var b = document.createHTMLElement(button); // b is HTMLButtonElement

// C# events
b.onclick += (s, e) => window.alert("Hi from C#!"); 

// Notice the use of LINQ
var divs = document.body.children.Where(x => x is HTMLDivElement);
```

@zydjohnHotmail The above project also has a Nuget  [Link for the Nuget](https://www.nuget.org/packages/WebView2.DOM/)

**Comment by zydjohnHotmail** [comment]
@darbid I have tried that repo., but it can't be installed on Windows 10, I think it works on Windows 7.  By the way, one of its dependency TypeConvert nuget version 2.1.6 is no longer in support.
In short, that repo has some issue, can not be easily used nowadays, at least for Windows 10.  I filed 2 issues on its repo., but after 3 days, there is no replies yet.

**Comment by darbid** [comment]
Lets hope they fix it.

**Comment by champnic** [comment]
Hey @genesislubrigas - thanks for the feature request, and sorry you hate WebView2 :( We are tracking this item in #77 (I'll update the title as we are tracking more than just Winforms) and as others have shared in this thread there are ways to do this today.

**Comment by genesislubrigas** [comment]
> Hey @genesislubrigas - thanks for the feature request, and sorry you hate WebView2 :( We are tracking this item in #77 (I'll update the title as we are tracking more than just Winforms) and as others have shared in this thread there are ways to do this today.

Yep, there are ways mentioned in this thread but why make it complicated in webview2 while DOM access with webbrowser was done easier.

I mean, if it can be done in webbrowser, why not put that feature as well in webview2 when we are already comfortable moving around with it.

**Comment by R2D221** [comment]
> @darbid I have tried that repo., but it can't be installed on Windows 10, I think it works on Windows 7. By the way, one of its dependency TypeConvert nuget version 2.1.6 is no longer in support.
> In short, that repo has some issue, can not be easily used nowadays, at least for Windows 10. I filed 2 issues on its repo., but after 3 days, there is no replies yet.

Hi, I'm the creator of the WebView2.DOM repo. It surprises me that you say that the project can't run on Windows 10 when all my development is done on Windows 10. If you would like te be more specific about what installations issues you're having, you can raise an issue in my repo. Do note that I'm just doing this for my pet projects, but if you give me some info I can help you run it on your side.

**Comment by amaitland** [comment]
Firstly sorry about the SPAM to those who are subscribed to multiple DOM issues in this repo. 

For anyone interested I've released [WebView2.DevTools.Dom](https://www.nuget.org/packages/WebView2.DevTools.Dom/) to [NuGet.org](https://www.nuget.org/profiles/amaitland).

It differs from WebView2.DOM in that is uses the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol) for communication. The [Microsoft.Web.WebView2.DevToolsProtocolExtension](https://www.nuget.org/packages/Microsoft.Web.WebView2.DevToolsProtocolExtension/) package is used for direct connection with the browser (doesn't open a remote debugging port).

- Ability to access/manipulate the DOM
- Can be used for automation, supports emulating keyboard and mouse events (like puppeteer)
- Basic Mobile device emulation support
- Emulate Keyboard/Mouse
- Supports returning of complex types from `JavaScript`
- Fully asynchronous 
- Supports multiple frames
- Approximately 280 `xUnit` tests

- [GitHub Page](https://github.com/ChromiumDotNet/WebView2.DevTools.Dom)

See https://github.com/MicrosoftEdge/WebView2Feedback/issues/77#issuecomment-1073677281 for some examples.

**Comment by aimcosoftware** [comment]
I have finally got round to finishing the Winforms wrapper with full sync DOM and API access.
Please see here: https://github.com/aimcosoftware/DOMView2
### Direct synchronous access to DOM and other APIs from .Net
```vb
'Strongly typed classes
Private WithEvents Btn As DVButton

'Insert dialog into the current page
Private Sub InsertDialog()
	Dim Dlg As DVDialog = DomView.document.createElement("dialog")
	Dlg.style.cssText = "box-shadow:5px 5px 10px grey;width:400px;border:1px solid silver;border-radius:10px;padding:10px"

	Dim Div As DVDiv = DomView.document.createElement("div")
	Div.style.cssText = "text-align:center;padding:20px"
	Div.innerText = "Dialog inserted from .Net"
	Dlg.insertAdjacentElement("afterBegin", Div)

	Btn = DomView.document.createElement("button")
	Btn.style.cssText = "border:1px solid silver;border-radius:5px;padding:10px 20px;float:right"
	Btn.innerText = "Close"
	Dlg.insertAdjacentElement("beforeEnd", Btn)

	DomView.document.body.insertAdjacentElement("beforeEnd", Dlg)
	Dlg.showModal()
End Sub

'Button click event
Private Sub Btn_AfterEvent(Sender As DVWindow, e As DVAfterEventArgs) Handles Btn.AfterEvent
	If e.type = "click" Then
		Dlg.close()
	End If
End Sub
```
